### PR TITLE
enhance autodoc_mock_imports

### PR DESCRIPTION
**Pull request by rjarry**
The `autodoc_mock_imports` option requires to explicitly declare _every_ external module and sub-module that are imported by the documented code. This is not practical as the list can become very large and must be maintained as the code changes.

Also, the mocking is minimal which causes errors when compiling the docs. For example, if you declare:

``` python
autodoc_mock_imports = ['jinja2.loaders']
```

And try to document a module:

``` rst
.. automodule:: my.lib.util
```

Which contains this code:

``` python
from jinja2.loaders import FileSystemLoader
LOADER = FileSystemLoader('/tmp')
```

The following error occurs:

```
  File ".../my/lib/util/templates.py" line 2
    LOADER = FileSystemLoader('/tmp')
TypeError: 'object' object is not callable
```

Other similar errors can occur such as `TypeError: 'object' object has no len`.

To address these limitations, only require to declare the top-level modules that should be mocked:

``` python
autodoc_mock_imports = ['jinja2']
```

Will mock `import jinja2` but also any sub-module: `import jinja2.loaders`, `from jinja2.ext import Stuff`, etc.

Also, make the mocked modules yield more complete dummy objects to avoid these `TypeError` problems.

### CODE REVIEW

**Comment by rjarry** [comment]
Hi,

I have 1 test that still fails:

```
$ make test
Checking dependencies...
Temporary files will be placed in .../sphinx/tests/build.
Running Sphinx test suite (with Python 2.7.9)...
................................S..................................................................................................................................................................................................................................................................................................................................................................................................................................................................SS....................................................................................................................F....................S....................................................................................................................................................................................................................................................S...................................................................
======================================================================
FAIL: test_ext_autosummary.test_get_items_summary
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../sphinx/.venv/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File ".../sphinx/tests/util.py", line 259, in deco
    func(app, status, warning, *args2, **kwargs2)
  File ".../sphinx/tests/test_ext_autosummary.py", line 83, in test_get_items_summary
    assert html_warnings == ''
AssertionError: 
    u'\x1b[31m.../sphinx/tests/build/test-autosummary/dummy_module.rst:4: WARNING: failed to import C.class_attr\n\x1b[39;49;00m\x1b[31m.../sphinx/tests/build/test-autosummary/dummy_module.rst:4: WARNING: failed to import C.prop_attr1\n\x1b[39;49;00m\x1b[31m.../sphinx/tests/build/test-autosummary/dummy_module.rst:4: WARNING: failed to import C.prop_attr2\n\x1b[39;49;00m\x1b[31m.../sphinx/tests/build/test-autosummary/dummy_module.rst:4: WARNING: failed to import C.C2\n\x1b[39;49;00m' = <StringIO.StringIO instance at 0x7fa0333d9d40>.getvalue()
>>  assert u'\x1b[31m.../sphinx/tests/build/test-autosummary/dummy_module.rst:4: WARNING: failed to import C.class_attr\n\x1b[39;49;00m\x1b[31m.../sphinx/tests/build/test-autosummary/dummy_module.rst:4: WARNING: failed to import C.prop_attr1\n\x1b[39;49;00m\x1b[31m.../sphinx/tests/build/test-autosummary/dummy_module.rst:4: WARNING: failed to import C.prop_attr2\n\x1b[39;49;00m\x1b[31m.../sphinx/tests/build/test-autosummary/dummy_module.rst:4: WARNING: failed to import C.C2\n\x1b[39;49;00m' == ''


----------------------------------------------------------------------
Ran 935 tests in 67.420s

FAILED (SKIP=5, failures=1)
Makefile:69: recipe for target 'test' failed
make: *** [test] Error 1
```

But when I run it separately, it works...

```
$ make test TEST=test_ext_autosummary
Checking dependencies...
Temporary files will be placed in .../sphinx/tests/build.
Running Sphinx test suite (with Python 2.7.9)...
..
----------------------------------------------------------------------
Ran 2 tests in 1.714s

OK
```

I am not sure on how to approach the problem :)

**Comment by robin-jarry** [comment]
Does anyone have an idea on how to fix this test error?

**Comment by rjarry** [comment]
I found where the problem was. I have changed the implementation a little bit. All tests should pass.

**Comment by stonier** [comment]
:+1: for this, could use it also.

**Comment by rjarry** [comment]
Hi,

I had forgotten about this :-) Is there something that prevents merging this patch ?

**Comment by tk0miya** [comment]
I guess nobody can review code around autodoc now...
I will try to review within a week (sorry, I'm in busy now).

**Comment by rjarry** [comment]
thanks!

Robin

**Comment by rjarry** [comment]
Hi @tk0miya

did you have time to look at the patch?

Let me know if there is anything I can do to help.

cheers

File: sphinx/ext/autodoc.py [hunk 1]
```diff
@@ -508,6 +513,36 @@ def parse_name(self):
...
+                        globals = {}
+                    if locals is None:
+                        locals = {}
+                    if level is None:
+                        level = -1
+                elif level is None:
+                    level = 0
+                return REALIMPORT(name, globals, locals, fromlist, level)
+            except ImportError:
+                if any(name.startswith(m)
```
Comment (tk0miya) [inline sphinx/ext/autodoc.py]: Is this correct? If I specified `autodoc_mock_imports = ['jinja']`, this might misdetect the module.

File: sphinx/ext/autodoc.py [hunk 2]
```diff
@@ -508,6 +513,36 @@ def parse_name(self):
...
+                if PY2:
+                    if globals is None:
+                        globals = {}
+                    if locals is None:
+                        locals = {}
+                    if level is None:
+                        level = -1
+                elif level is None:
+                    level = 0
+                return REALIMPORT(name, globals, locals, fromlist, level)
```
Comment (tk0miya) [inline sphinx/ext/autodoc.py]: So far, the modules in `autodoc_mock_imports` are not imported. But it seems this try to import the module once. Is this change intentional?

File: sphinx/ext/autodoc.py [hunk 3]
```diff
@@ -508,6 +513,36 @@ def parse_name(self):
...
+            except ImportError:
+                if any(name.startswith(m)
+                       for m in self.env.config.autodoc_mock_imports):
+                    self.env.app.debug('[autodoc] adding a mock module %s!'
+                                       % name)
+                    return _MockModule(name)
+                else:
+                    raise
+
+        builtins.__import__ = _import
```
Comment (tk0miya) [inline sphinx/ext/autodoc.py]: overriding `builtins` is well-known way to hook importing?
(Sorry, I don't know the mechanism of loading modules...)

**Review by tk0miya — CHANGES_REQUESTED** [review]
LGTM with nits

**Comment by tk0miya** [comment]
Sorry for late response. I reviewed now.

File: sphinx/ext/autodoc.py (hunk 2 above)
Comment (rjarry) [reply sphinx/ext/autodoc.py]: Yes, if a module is available, we should use it instead of return a mocked one.

File: sphinx/ext/autodoc.py (hunk 1 above)
Comment (rjarry) [reply sphinx/ext/autodoc.py]: No, since jinja is a sphinx dependency, it should be available in sys.path and `REALIMPORT` will return the real module, not a mocked one.

File: sphinx/ext/autodoc.py (hunk 3 above)
Comment (rjarry) [reply sphinx/ext/autodoc.py]: I am not sure this is a well know way to do it, but it is the best I found to cover all problems.

Also, I like this approach since it is lazy. A module will be mocked only when required.

File: sphinx/ext/autodoc.py (hunk 1 above)
Comment (rjarry) [reply sphinx/ext/autodoc.py]: If you want I could make sure that all mocked imports are cleaned up when done processing the autodoc.

**Comment by rjarry** [comment]
I'll fix conflicts tonight and add some code to clean up mocked imports when after autodoc is finished.

File: sphinx/ext/autodoc.py (hunk 3 above)
Comment (shimizukawa) [reply sphinx/ext/autodoc.py]: Please use import hook mechanism: https://docs.python.org/3/reference/import.html

Before python-2.3, we have to replace `__import__` method to hook importing mechanism. However [PEP-0302](https://www.python.org/dev/peps/pep-0302/) provides a way to hook importing.
[PEP-0451](https://www.python.org/dev/peps/pep-0451/) is latest specification of importing, however (I think) PEP-0451 is not applied to Python-2.7.

My conclusion is, we can implement the purpose of this PR with using PEP-0302 spec for both of Py2 and Py3.

File: sphinx/ext/autodoc.py (hunk 2 above)
Comment (tk0miya) [reply sphinx/ext/autodoc.py]: Okay, I think it is an incompatible change. Please update the document.

@shimizukawa Is this change acceptable? (I'm not a user of autodoc, so I can't say this is okay or not...)

File: sphinx/ext/autodoc.py (hunk 1 above)
Comment (tk0miya) [reply sphinx/ext/autodoc.py]: Ah, sorry. my question was poor. In this change, you use `startswith()` to compare the module names. so I worry about `autodoc_mock_imports = ['foo']` matches `foobar` module. Is this right?

File: sphinx/ext/autodoc.py (hunk 3 above)
Comment (rjarry) [reply sphinx/ext/autodoc.py]: Hi @shimizukawa, I did not know about import hooks. I'll try to understand how it works and update the PR.

File: sphinx/ext/autodoc.py (hunk 1 above)
Comment (rjarry) [reply sphinx/ext/autodoc.py]: Yes you are right. Maybe the matching should be more restrictive.

**Comment by rjarry** [comment]
@tk0miya, @shimizukawa, I have updated the patch

it now uses import hooks as specified in [PEP-0302](https://www.python.org/dev/peps/pep-0302/) and the mocked modules are removed from `sys.modules` to avoid side-effects in code that runs after autodoc.

File: sphinx/ext/autodoc.py [hunk 4]
```diff
@@ -107,49 +106,96 @@ def __getattr__(self, name):
...
+
+    def __init__(self, names):
+        # type: (List[str]) -> None
+        self.packages = set()  # type: Set[str]
+        for n in names:
+            # Convert the list of module names:
+            #     ['a.b.c', 'd.e']
+            # to a set of base packages:
+            #     set(['a', 'd'])
+            self.packages.add(n.split('.')[0])
```
Comment (shimizukawa) [inline sphinx/ext/autodoc.py]: IMPORTANT: How do we mock under `xml.etree` module except `xml.dom` ?
IMO, it makes confusion that this _MockImportHook will mock `xml` and all under the module even if user specify `xml.etree`.

File: sphinx/ext/autodoc.py [hunk 5]
```diff
@@ -107,49 +106,96 @@ def __getattr__(self, name):
...
+        self.__all__ = []  # type: List[str]
+
+    def __getattr__(self, name):
+        # type: (str) -> _MockObject
+        o = _MockObject()
+        o.__module__ = self.__name__
+        return o
+
+
+class _MockImportHook(object):
```
Comment (shimizukawa) [inline sphinx/ext/autodoc.py]: nits: It is better to change the class name to MockLoader according to other loaders.

**Review by shimizukawa — CHANGES_REQUESTED** [review]
Reviewed. Thanks for your updating the code! 
I am very glad that the implementation has been improved to a very simple implementation.

I've left a comment about the part I was concerned about.

File: sphinx/ext/autodoc.py (hunk 5 above)
Comment (rjarry) [reply sphinx/ext/autodoc.py]: Well, it is technically both a module finder and loader. That's why I did not call it "loader". Maybe I can rename it to `_MockImporter`?

I do not have a strong opinion about it though. If you want I can rename it to `_MockLoader`.

File: sphinx/ext/autodoc.py (hunk 4 above)
Comment (rjarry) [reply sphinx/ext/autodoc.py]: Yeah I understand your concern. The problem is that in order to mock `a.b`, you need to also mock `a`. This is explained in [PEP 302](https://www.python.org/dev/peps/pep-0302/):

> The Importer Protocol operates at this level of individual imports. By the time an importer gets a request for "spam.ham", module "spam" has already been imported.

If you want, I can do a more complex algorithm to make sure that if `autodoc_mock_imports = ['a.b']`, only `a`, `a.b` and sub packages (`a.b.*`) are mocked and not `a.c`.

But since I remove the mocked modules from `sys.modules` after running the auto documenter, this should not have any consequences in Sphinx overall.

Comment (rjarry) [reply sphinx/ext/autodoc.py]: Also, if I understand the motivation of this feature, this is to be able to run autodoc on code that depends on third party libraries, without requiring to install them when generating the doc.

This means that in most cases, users will need to mock: `django`, `paramiko`, `twisted` and any of their respective sub-packages. I don't see a concrete use-cases where you would only need to mock parts of third-party libraries (for example: mock `django.db` and not `django.contrib`).

**Comment by rjarry** [comment]
@shimizukawa, I have updated the patch following your remarks.

File: sphinx/ext/autodoc.py (hunk 5 above)
Comment (shimizukawa) [reply sphinx/ext/autodoc.py]: `_MockImporter` is fine to me ;)

File: sphinx/ext/autodoc.py (hunk 4 above)
Comment (shimizukawa) [reply sphinx/ext/autodoc.py]: Well, you are right. We should avoid "complex algorithm" and should keep simple specification.
I agree to your *simple* implementation.

c1761c1 looks *complex* one and it needs long note, so could you revert it?
I'm sorry to trouble you.

**Review by shimizukawa — APPROVED** [review]

**Comment by shimizukawa** [comment]
LGTM with nits: please refer https://github.com/sphinx-doc/sphinx/pull/2961#discussion_r111965033

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
LGTM!

File: sphinx/ext/autodoc.py (hunk 4 above)
Comment (rjarry) [reply sphinx/ext/autodoc.py]: No problem, I'll revert to my previous implementation.

**Comment by rjarry** [comment]
I reverted to a simple algorithm to determine which modules to mock. It should be ok now :)

**Comment by shimizukawa** [comment]
@rjarry Thanks!!

**Comment by rjarry** [comment]
Thanks for the thorough review @shimizukawa @tk0miya :-)

**Comment by TTRh** [comment]
I may be wrong but seems that with this PR, autodoc now skip subclasses of object from mocked module. For example : 

my_module.py

```
from foo import Bar

class MyClass(Bar)
    """"This is my class""""
```

my_module.rst
```
.. autoclass:: MyClass
    :members:
```

conf.py

```
autodoc_mock_imports = ['foo']
```
It seems that with this version the extension sees now `MyClass` as mocked class ...
